tensorflow/compiler/xla/service/gpu/fusion_bitcast_lift.cc

```diff
@@ -0,0 +1,205 @@
+      return false;
+    }
+  }
+  return true;
+}
+
+StatusOr<bool> FusionBitcastLift::Run(HloModule* module) {
+  XLA_VLOG_LINES(2, "FusionBitcastLift::Run(), before:\n" + module->ToString());
+  bool changed = false;
+  for (HloComputation* comp : module->MakeNonfusionComputations()) {
+    // Copy the instruction list as we modify the HloComputation.
+    std::vector<HloInstruction*> comp_instruction(comp->instructions().begin(),
+                                                  comp->instructions().end());
+    for (auto it = comp_instruction.begin(); it != comp_instruction.end();
+         it++) {
+      // 1) Is this a fusion that we want to modify.
+      HloInstruction* instr = *it;
+      if (HloFusionInstruction* fusion = DynCast<HloFusionInstruction>(instr)) {
+        // 1.1) We only support kInput fusion and some operations.
+        if (fusion->fusion_kind() != HloInstruction::FusionKind::kInput ||
+	    !AreInstructionSupported(
+                fusion->fused_instructions_computation())) {
+          continue;
+        }
+        // 1.2) Check if there is a bitcast that we lift. Currently
+        //      we do not lift(merge) bitcast above(with) broadcast.
+        if (!std::any_of(fusion->fused_instructions().begin(),
+                         fusion->fused_instructions().end(),
+                         [](HloInstruction* inner) {
+              return inner->opcode() == HloOpcode::kBitcast &&
+                  inner->operand(0)->opcode() != HloOpcode::kBroadcast;
+                         })) {
+          continue;
+        }
+        // 1.3) Check that all the bitcast have the same shape pattern.
+        //      Multiple bitcast pattern isn't supported/tested.
+	std::vector<HloInstruction*> bitcasts;
+	for (HloInstruction* fused_instr : fusion->fused_instructions()) {
+	  if (fused_instr->opcode() == HloOpcode::kBitcast &&
+	      fused_instr->shape().rank() <
+	      fused_instr->operand(0)->shape().rank()) {
+	    if (bitcasts.size() > 0 && (
+                    !ShapeUtil::Equal(fused_instr->shape(),
+				      bitcasts[0]->shape()) ||
+                    !ShapeUtil::Equal(bitcasts[0]->operand(0)->shape(),
+				      fused_instr->operand(0)->shape()))) {
+	      continue;
+	    }
+	    bitcasts.push_back(fused_instr);
+	  }
+	}
+
+	// 2) Now that we have found a fusion that we want to modify,
+	//    create the new fusion. We do so by:
+	//    a) Cloning the old fusion.
+	//    b) Recursively walk the graph from the root and lift the
+	//       bitcast up across one instruction at a time.
+	std::unique_ptr<HloInstruction> cloned_fusion =
+	  fusion->Clone("bitcast");
+	std::vector<HloInstruction*> stack(
+					   {cloned_fusion->fused_expression_root()});
+	bool clone_changed = false;
+	while (stack.size() > 0) {
+	  HloInstruction* i = stack.back();
+	  stack.pop_back();
+	  if (i->opcode() == HloOpcode::kTuple) {
+	    stack.insert(stack.end(), i->operands().begin(),
+			 i->operands().end());
+	    continue;
```

Comment:
Also, it looks to me that the "continue;" is not needed.

Reply:
I did all the changes expect this one. I have difficulty making an example that demonstrate this.
Can you share 2 hlo line: the constant and the bitcast of the constant? With that, I'll be sure to add the extra optimization that you want to see.

Reply:
So essentially it was a sequence like this:
bitcast(add(constant, parameter))
All users of the add are bitcasts, so the bitcast is lifted, and the shape of add is adjusted. Then, parameter is adjusted as well, but constant is not adjusted, which makes the HLO invalid because add expects the two operands to have the same shape.
The shapes involved in the bitcast were [1] and [] (i.e. casting from scalar to 1-element tensor).
Sorry, I forgot to copy the exact HLO, but I think with this additional information you should be able to create a test case out of it.

Reply:
Thinking about this a bit more, I believe there is a bigger problem hidden here:
The algorithm assumes that if all users of an op are (the same type of) bitcasts, it is always possible to lift it. But it also depends on the users of the bitcasts whether this is possible, e.g. if the user of a bitcast is not a unary op, but e.g. a binary op that expects equal shapes as operands, it will require that both its operands will be changed at the same time. The handling of constants does not fix this general issue, it would just make this particular test work.
So in addition to lifting the bitcast, you would have to look at all users of the bitcast whether they need additional bitcasts inserted so that their "equal shapes" constraint is still satisfied. It becomes pretty complicated at that point. Or you just allow lifting bitcasts in case there are only unary ops as users of the bitcasts (but that might be too limiting).